prusti-encoder/src/encoders/mir_impure.rs

```diff
@@ -816,8 +820,56 @@ impl<'vir, 'enc, E: TaskEncoder> mir::visit::Visitor<'vir> for ImpureEncVisitor<
                 ..
             } => {
                 let (func_def_id, caller_substs) = self.get_def_id_and_caller_substs(func);
-                let is_pure = crate::encoders::with_proc_spec(func_def_id, |spec|
-                    spec.kind.is_pure().unwrap_or_default()
+
+                // Locate the impl for any trait bounds on the callee (w.r.t to
+                // input args), and encode the appropriate axiom(s). For example
+                // suppose we are calling f<T: Copy>(t: T) via a call f(bool).
+                // Then, we need to ensure there is an axiom asserting
+                // `implements_Copy(s_Bool_type())`.
+                let constraints = self.vcx
+                    .tcx()
+                    .predicates_of(func_def_id)
+                    .instantiate(self.vcx.tcx(), caller_substs);
+
+                let infcx = self.vcx.tcx().infer_ctxt().build();
+
+                for constraint in constraints.predicates {
+                    match constraint.as_predicate().kind().no_bound_vars().unwrap() {
+                        PredicateKind::Clause(ClauseKind::Trait(trait_predicate)) => {
+                            let obligation = Obligation {
+                                cause: ObligationCause::dummy(),
+                                param_env: self.vcx.tcx().param_env(self.def_id),
+                                predicate: trait_predicate,
+                                recursion_depth: 0,
+                            };
+                            let mut selcx = SelectionContext::new(&infcx);
+                            match selcx.select(&obligation) {
+                                Ok(Some(sel)) => match sel {
+                                    traits::ImplSource::UserDefined(ud) => {
+                                        self.deps.require_dep::<UserDefinedTraitImplEnc>(ud.impl_def_id).unwrap();
+                                    }
+                                    traits::ImplSource::Param(_) => {}
```

Comment:
I'm a little surprised to see that there's nothing generated for trait bounds on parameters? Don't we need to generate some opaque value so that we can instantiate trait axioms?

Reply:
I'm afraid I don't quite understand this comment. I think what the `ImplSource::Param` case corresponds to is e.g. the call to `g(x)` in :

```rust
fn f<T: Clone + Eq>(x: T) {
    g(x)
}

fn g<T: Clone>(t: T) {

}
```

where the call is allowed because the type of `x` must implement `Clone` in `f`. So we don't need to add any additional Viper axioms because the precondition `implements_Clone(typeof(x))` in the encoding of `f` will satisfy the corresponding precondition in `g`.

BTW in case it's not clear this would be different for a call of e.g. `g(1)` somewhere in the code; we would want to ensure the axiom asserting `implements_Clone(s_u32_type())` is generated.

Reply:
ah yea, i didn't catch that this was for the bounds of a called function, where indeed we don't want to generate axioms.